ocr.py

```diff
@@ -0,0 +1,174 @@
+                draw.text(top_left, text, fill=text_color)
+            except Exception as e:
+                logging.error(f"Error drawing rectangle - {top_left} - {bottom_right}: {e}")
+        return drawable_image
+
+    def filter_ocr_result(self, result, exclude_keywords=['retroarch', 'retronrch']):
+        """
+        Filter out unwanted OCR results based on keywords.
+
+        :param result: OCR result containing bounding boxes and text
+        :param exclude_keywords: List of keywords to filter out
+        :return: Filtered OCR result
+        """
+        return [(bbox, text, prob) for bbox, text, prob in result if not any(keyword in text.lower() for keyword in exclude_keywords)]
+
+    def ocr_easyocr(self, image_bytes, detail=1):
+        """
+        Perform OCR using easyocr.
+
+        :param image_bytes: The image in bytes
+        :param detail: Level of detail for OCR results
+        :return: OCR result containing bounding boxes and text
+        """
+        if self.method == 1:
+            return self.reader.readtext(image_bytes, detail=detail)
+    
+    def det_easyocr(self, image_bytes):
+        """
+        Perform text detection using easyocr.
+
+        :param image_bytes: The image in byes
+        :return:OCR result containing bounding boxes
+        """
+        return self.reformat(self.reader.detect(image_bytes))
+    
+    def ocr_openai(self, image_bytes):
+        response = self.openai_api.call_vision_api(image_bytes)
+        content = response['choices'][0]['message']['content']
+        return content
+
+
+    def ocr_and_highlight(self, image):
+        """
+        Perform OCR on the image and highlight the detected text.
+
+        :param image: The input PIL Image
+        :return: Tuple containing the concatenated detected text, annotated image, and OCR result
+        """
+        image_bytes = self.process_image(image)
+        if self.method == 1:
+            result = self.ocr_easyocr(image_bytes, detail=1)
+            filtered_result = self.filter_ocr_result(result)
+            drawable_image = self.draw_highlight(image_bytes, filtered_result)
+            filtered_text = ' '.join([text for _, text, _ in filtered_result])
+            return filtered_text, drawable_image, filtered_result
+        elif self.method == 2:
+            detection_result = self.det_easyocr(image_bytes)
+            dialogue_box_img = image_crop_dialogue_box(image, detection_result)
+            dialogue_box_image_bytes = self.process_image(dialogue_box_img)
+            drawable_image = self.draw_highlight(image_bytes, detection_result)
+            ocr_result = self.ocr_openai(dialogue_box_image_bytes)
+            if self.lang == 'en':
+                filtered_text = ocr_result.removeprefix('The text in the photo reads:').removeprefix('The text in the photo says:').replace('`', '').replace('\n', ' ').replace('"', '').strip()
```

Comment:
lets leave a todo here, we should probably fix the initial openai prompts instead of just removing prefix, we can do this later but put a note in the code why this weird thing is here